libraries/ms-common/inc/critical_section.h

```diff
@@ -14,10 +14,23 @@
 // This will also protect nested attempts from enabling and disabling interrupts from prematurely
 // ending the critical section.
 
+// Use this macro to mark a function as critical - it will automatically be ended as the function
+// goes out of scope.
+// EXAMPLE:
+//
+// void foo(void) {
+//   CRITICAL_SECTION_AUTOEND;
+//   // Critical code here.
+// }
+#define CRITICAL_SECTION_AUTOEND \
+  __attribute__((cleanup(_critical_section_cleanup))) bool _disabled = critical_section_start();
```

Comment:
Please double check this compiles with clang. I'd like to be able to keep using ASAN and TSAN. Commands if you want to be lazy.

```bash
make reallyclean
make build_all PLATFORM=x86 COMPILER=clang
```

Reply:
There are a bunch of problems with other modules, but the cleanup works with clang. I'll include the fixes to make clang build in this PR.